lib/backend/k8s/resources/ippools.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2016 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s/thirdparty"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	"k8s.io/client-go/kubernetes"
+	kerrors "k8s.io/client-go/pkg/api/errors"
+	kapiv1 "k8s.io/client-go/pkg/api/v1"
+	extensions "k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/rest"
+)
+
+func NewIPPools(c *kubernetes.Clientset, r *rest.RESTClient) api.Client {
+	return &client{
+		clientSet: c,
+		tprClient: r,
+	}
+}
+
+// Implements the api.Client interface for pools.
+type client struct {
+	clientSet *kubernetes.Clientset
+	tprClient *rest.RESTClient
+}
+
+func (c *client) Create(kvp *model.KVPair) (*model.KVPair, error) {
+	tpr := IPPoolToThirdParty(kvp)
+	res := thirdparty.IpPool{}
+	req := c.tprClient.Post().
+		Resource("ippools").
+		Namespace("kube-system").
+		Body(tpr)
+	err := req.Do().Into(&res)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, kvp.Key)
+	}
+	kvp.Revision = res.Metadata.ResourceVersion
+	return kvp, nil
+}
+
+func (c *client) Update(kvp *model.KVPair) (*model.KVPair, error) {
+	tpr := IPPoolToThirdParty(kvp)
+	res := thirdparty.IpPool{}
+	req := c.tprClient.Put().
+		Resource("ippools").
+		Namespace("kube-system").
+		Body(tpr).
+		Name(tpr.Metadata.Name)
+	err := req.Do().Into(&res)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, kvp.Key)
+	}
+	kvp.Revision = tpr.Metadata.ResourceVersion
+	return kvp, nil
+}
+
+func (c *client) Apply(kvp *model.KVPair) (*model.KVPair, error) {
+	updated, err := c.Update(kvp)
+	if err != nil {
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			return nil, err
+		}
+
+		// It doesn't exist - create it.
+		updated, err = c.Create(kvp)
```

Comment:
Is there no equivalent to Apply in the k8s client?

Also, for future, is there a way to specify revision numbers to prevent concurrent updates - we'll want to expose that in the API eventually.